fontmaker/core/ttfparser.go

```diff
@@ -204,65 +204,69 @@ func (t *TTFParser) SetUseKerning(use bool) {
 }
 
 //Parse parse
-func (t *TTFParser) Parse(filepath string) error {
+func (t *TTFParser) Parse(filepath string) ([]byte, error) {
 	data, err := ioutil.ReadFile(filepath)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	buff := bytes.NewBuffer(data)
-	return t.parse(buff)
+	return t.parse(buff, nil)
 }
 
 //ParseByReader parse by io.reader
-func (t *TTFParser) ParseByReader(rd io.Reader) error {
-	return t.parse(rd)
+func (t *TTFParser) ParseByReader(rd io.Reader, fontdata []byte) ([]byte, error) {
+	return t.parse(rd, fontdata)
 }
 
-func (t *TTFParser) parse(rd io.Reader) error {
+func (t *TTFParser) parse(rd io.Reader, loadedFontdata []byte) ([]byte, error) {
 
-	fontdata, err := ioutil.ReadAll(rd)
-	if err != nil {
-		return err
+	var fontdata []byte
+	if loadedFontdata == nil {
+		var err error
+		fontdata, err = ioutil.ReadAll(rd)
```

Comment:
According to CPU profiling, here is the most expensive process. This patch allows us to avoid it when the loaded font data is provided to `parse()`.

Reply:
If this analysis is correct, then what we want to have is a method that we don't need to call `ReadAll()`, say

```go
func (t *TTFParser) ParseByBytesReader(fr *bytes.Reader) error {
   // does stuff after `fd := bytes.NewReader(fontdata)` line
}
```

then we would want to call this function with a `*bytes.Reader` with a pre-loaded font bytes.

Reply:
📝  We should probably be using `io.ReadAll`? Or are we supporting older versions of Go?

Also, `ReadAll()` just calls the `Reader`'s `Read()` method: https://golang.org/src/io/io.go?s=19435:19470#L633
You can see in Toshi's previous analysis that the expensive time occurs in the underling tar Reader's `Read()` method (https://golang.org/pkg/archive/tar/#Reader.Read), which (as far as I can tell) is reading from a file.
![image](https://user-images.githubusercontent.com/10292869/113211955-a2f9cf00-92d2-11eb-8c75-61679083693c.png)

A bytes reader however (https://golang.org/src/bytes/reader.go?s=1154:1204#L30) basically just performs an in-memory copy of the data from one buffer to another, which should be significantly faster I think. 

Let's do a proper analysis, but we shouldn't conclude that we should avoid `ReadAll()` yet I don't think

Reply:
I was trying to use `TeeReader` to check its performance but am not sure how it is supposed to be used around `ParseFontData()`. 🤔